waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs

```diff
@@ -0,0 +1,125 @@
+-- We make it pending instead, and once the new message comes from the same job, we apply it at the
+-- start of that message.
+--
+-- This way we get proper output in the situation as descried above:
+--   Server: First line
+--   Some other output!
+--   Server: Second line
+printPrefixedJobMessage :: J.JobMessage -> PrefixedWriter ()
+printPrefixedJobMessage jobMessage = do
+  jobsWithPendingNewline <- gets _jobsWithPendingNewline
+  let (jobsWithPendingNewline', content') =
+        applyPendingNewline jobsWithPendingNewline (jobType, content)
+  modify $ \s -> s {_jobsWithPendingNewline = jobsWithPendingNewline'}
+
+  liftIO $ do
+    T.IO.hPutStr outHandle $ addPrefixWhereNeeded content'
+    hFlush outHandle
+  where
+    addPrefixWhereNeeded :: T.Text -> T.Text
+    addPrefixWhereNeeded =
+      let prefix = makeJobMessagePrefix jobMessage
+          addPrefixAfterSubstr substr = T.intercalate (substr <> prefix) . T.splitOn substr
+       in addPrefixAfterSubstr "\r" . addPrefixAfterSubstr "\n"
+
+    outHandle = getJobMessageOutHandle jobMessage
+    jobType = J._jobType jobMessage
+    content = getJobMessageContent jobMessage
+
+newtype PrefixedWriter a = PrefixedWriter {_runPrefixedWriter :: StateT PrefixedWriterState IO a}
+  deriving (Functor, Applicative, Monad, MonadIO, MonadState PrefixedWriterState)
+
+data PrefixedWriterState = PrefixedWriterState
+  {_jobsWithPendingNewline :: JobsWithPendingNewline}
+
+runPrefixedWriter :: PrefixedWriter a -> IO a
+runPrefixedWriter pw = fst <$> runStateT (_runPrefixedWriter pw) initState
+  where
+    initState =
+      PrefixedWriterState
+        { _jobsWithPendingNewline = S.fromList [minBound ..]
+        }
+
+type JobsWithPendingNewline = S.Set J.JobType
+
+-- | Given a set of jobs with pending newline and a job message,
+-- it applies any pending newline (newline from the previous messages of the same job type)
+-- to the job message content while also detecting if content ends with a newline
+-- and in that case adds it to the set of pending newlines (while removing used pending newline).
+-- It returns this updated content and updated set of pending newlines.
+applyPendingNewline ::
+  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
+applyPendingNewline jobsWithPendingNewline (jobType, content) =
+  (jobsWithPendingNewlines', content')
+  where
+    content' =
+      let addPendingNewlineToStartIfAny =
+            if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
+          removeTrailingNewlineIfAny =
+            if contentEndsWithNewline then T.init else id
+       in addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
+    jobsWithPendingNewlines' =
+      if contentEndsWithNewline
+        then S.insert jobType jobsWithPendingNewline
+        else S.delete jobType jobsWithPendingNewline
+    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

Comment:
I found some of the inner "if-ing" a bit hard to grok... is this any easier perhaps?

```haskell
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) =
  let content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      jobsWithPendingNewlines' = updateOp jobType jobsWithPendingNewline
   in (jobsWithPendingNewlines', content')
  where
    updateOp :: JobType -> Set JobType -> Set JobType
    updateOp = if contentEndsWithNewline then S.insert else S.delete

    addPendingNewlineToStartIfAny = if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    removeTrailingNewlineIfAny = if contentEndsWithNewline then T.init else id
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

Reply:
I agree this part is a bit complex, I tried simplifying it couple of times but never got to perfect solution. This looks interesting I will try using it, maybe doing some hybrid!

Reply:
I went with something in between and ended up with this:
```hs
applyPendingNewline ::
  JobsWithPendingNewline -> (J.JobType, T.Text) -> (JobsWithPendingNewline, T.Text)
applyPendingNewline jobsWithPendingNewline (jobType, content) = (jobsWithPendingNewline', content')
  where
    content' = addPendingNewlineToStartIfAny $ removeTrailingNewlineIfAny content
      where
        removeTrailingNewlineIfAny =
          if contentEndsWithNewline then T.init else id
        addPendingNewlineToStartIfAny =
          if jobType `S.member` jobsWithPendingNewline then ("\n" <>) else id
    jobsWithPendingNewline' = updateOp jobType jobsWithPendingNewline
      where
        updateOp = if contentEndsWithNewline then S.insert else S.delete
    contentEndsWithNewline = "\n" `T.isSuffixOf` content
```

What does it look like?

Reply:
Looks good to me!